circuit/src/merkle_tree/native.rs

```diff
@@ -0,0 +1,166 @@
+use std::marker::PhantomData;
+
+use halo2wrong::halo2::arithmetic::FieldExt;
+use num_traits::pow;
+
+use crate::{
+	params::RoundParams,
+	poseidon::native::{sponge::PoseidonSponge, Poseidon},
+};
+const WIDTH: usize = 5;
+
+#[derive(Clone, Debug)]
+struct MerkleTree<F: FieldExt, P>
+where
+	P: RoundParams<F, WIDTH>,
+{
+	_params: PhantomData<P>,
+	root: Option<F>,
+	nodes: Vec<Option<F>>,
+	height: usize,
+	first_leaf_index: usize,
+}
+
+#[derive(Clone)]
+struct Path<F: FieldExt, P>
+where
+	P: RoundParams<F, WIDTH>,
+{
+	_params: PhantomData<P>,
+	value: F,
+	path_vec: Vec<F>,
+}
+
+impl<F: FieldExt, P> MerkleTree<F, P>
+where
+	P: RoundParams<F, WIDTH>,
+{
+	fn new(height: usize) -> Self {
+		// An array starts from index 0, that is why we do -1 operation to find the
+		// first leaf index.
+		let first_leaf_index = pow(2, height) - 1;
+		// Calculating number of nodes is basically 2n - 1 (n is number of leaves)
+		let num_nodes = pow(2, height + 1) - 1;
+		let nodes = vec![None; num_nodes];
+		MerkleTree { _params: PhantomData, root: None, nodes, height, first_leaf_index }
+	}
+
+	fn add_leaves(&mut self, values: Vec<F>) -> &Self {
+		for i in 0..values.len() {
+			self.nodes[self.first_leaf_index + i] = Some(values[i]);
+		}
+		self
+	}
+
+	fn build_tree(&mut self) -> &Self {
+		if self.height > 0 {
+			let mut new_tree = MerkleTree::<F, P>::new(self.height - 1);
+			let first_leaf_index = (new_tree.first_leaf_index * 2) + 1;
+			let mut hash = Vec::new();
+			for i in 0..first_leaf_index {
+				if i % 2 == 0 {
+					let hasher: Poseidon<F, WIDTH, P> = Poseidon::new([
+						self.nodes[first_leaf_index + i].unwrap(),
+						self.nodes[first_leaf_index + i + 1].unwrap(),
+						F::zero(),
+						F::zero(),
+						F::zero(),
+					]);
+					let hashes = hasher.permute();
+
+					hash.push(hashes[0] + hashes[1]);
```

Comment:
Just curious, why are we adding hash[0] and hash[1]?

Reply:
I wanted to concatenate both hash values by adding them.